api/httpclient.go

```diff
@@ -0,0 +1,183 @@
+package api
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httputil"
+	"strings"
+	"time"
+
+	"github.com/decred/dcrd/chaincfg/v3"
+)
+
+const (
+	// Default http client timeout in secs.
+	defaultHttpClientTimeout = 30 * time.Second
+)
+
+// Client is the base for http calls
+type (
+	Client struct {
+		httpClient *http.Client
+		Debug      bool
+		BaseUrl    string
+		ReqFilter  RequestFilter
+	}
+	RequestFilter func(info RequestInfo) (req *http.Request, err error)
+
+	// ClientConf Models http client configurations
+	ClientConf struct {
+		Debug   bool
+		BaseUrl string
+	}
+
+	// RequestInfo models the http request data.
+	RequestInfo struct {
+		client  *Client
+		request *http.Request
+		Payload interface{}
+		Method  string
+		Url     string
+	}
+)
+
+// NewClient return a new HTTP client
+func NewClient(conf *ClientConf) (c *Client) {
+	t := http.DefaultTransport.(*http.Transport).Clone()
+	t.TLSClientConfig = &tls.Config{
+		InsecureSkipVerify: true,
+	}
+
+	client := &http.Client{
+		Timeout:   defaultHttpClientTimeout,
+		Transport: t,
+	}
+
+	return &Client{
+		httpClient: client,
+		Debug:      conf.Debug,
+		BaseUrl:    conf.BaseUrl,
+		ReqFilter:  nil,
+	}
+}
+
+// doTimeoutRequest sends a HTTP request with a timeout
+func (c *Client) doTimeoutRequest(timer *time.Timer, req *http.Request) (*http.Response, error) {
+	// Send the request in the background
+	type result struct {
+		resp *http.Response
+		err  error
+	}
+	done := make(chan result, 1)
+	go func() {
+		c.dumpRequest(req)
+		resp, err := c.httpClient.Do(req)
+		c.dumpResponse(resp)
+		done <- result{resp, err}
+	}()
+	// Wait for the read or timeout
+	select {
+	case r := <-done:
+		return r.resp, r.err
+	case <-timer.C:
+		errTimeout := errors.New("timeout reading from API")
+		return nil, errTimeout
+	}
```

Comment:
I think http request with timeout could be made using `context.WithTimeout` (and using this ctx with the actual request you are making), that will basically replace the need for `doTimeoutRequest` method